[ADD] voip_oca

etobella opened the pull request:
This module allows the use of VOIP directly from Odoo.

It relies on SIP.js to connect to the PBX using a websocket directly from the browser.

Odoo server will not connect directly to the PBX server, but will have users and passwords stored.

Credits
======

The development of this module has been financially supported by:

- Associacion Española de Odoo ([AEODOO](https://www.aeodoo.org/))

Configuration
=============

Create the PBX Connection
-------------------------

-  Access in Debug mode.
-  Go to ``Settings > Technical > Discuss > PBX Servers``.
-  Create a PBX server and define the domain name and websocket link.

You can set it as ``Development`` or ``Production``. Production environment will never contact the PBX server.

Configure users
---------------

For each user, we need to define their PBX server, user and password. To do this, we have two options:

1. Configure users as an admin and define the information directly 
2. As a user, go to your Preferences and modify the information.

Usage
=====

Automatically the system will login into the PBX server once we enter odoo with our user. If we are on a test environment, no login is made and the libraries of SIP are not loaded.

Once we are logged, we can use the VOIP widget in order to call directly. In order to open all the selectors, we can press on the top bar and the bottom widget will be opened.

![image](https://github.com/user-attachments/assets/ee5aa5c1-497c-4d56-ae2e-17ab5d5da523)

We can make call by:

1. Select the partner and press the number
2. Select an activity and press the number of the associated partner
3. Select an old call and press the number
4. Press the keyboard and write the number directly

![image](https://github.com/user-attachments/assets/255aa76d-a55d-47a3-ba32-86d02ec8b220)

Inside a call, we can transfer the call, end the call, mute the call and put the call on Hold.

Also, the system allows to receive call. In that case, the system will try to find the related partner and will open the widget automatically.

![image](https://github.com/user-attachments/assets/12032531-1e73-49ae-8c9d-980b57d03538)

Known issues / Roadmap
======================

The following options are interesting but are not implemented right now:

-  Allow to enable or disable the VOIP as a user (Login / Logout)
-  Create an autocall system where the calls are done automatically according to tasks

The following options will be done before release:

- [x]  Show the status of the VOIP connection on the header
- [x] Allow to call from a phone widget

etobella committed:
[ADD] voip_oca

Co-authored-by: Luis Rodríguez <luis.rodriguez@dixmit.com>

alexis-via commented on voip_oca/README.rst:
```diff
@@ -0,0 +1,171 @@
+-  Go to ``Settings > Technical > Discuss > PBX Servers``.
+-  Create a PBX server and define the domain name and websocket link.
+
+You can set it as ``Development`` or ``Production``. Production
```
At the end of line: Production => Development

JordiBForgeFlow commented on voip_oca/models/mail_activity.py:
```diff
@@ -0,0 +1,44 @@
+class VoipOcaActivity(models.Model):
+    _inherit = "mail.activity"
+
+    main_partner_id = fields.Many2one(
```
I would suggest to make the module dependent on mail_activity_partner https://github.com/OCA/social/pull/1606 instead.

pedrobaeza replied:
Please don't. Base modules shouldn't depend on others not base one. I think the activity should be put to the user, not the partner.

etobella replied:
But we need the partner to show the partner that we are going to call. That's why we used this.

The problem with the module that you are suggesting is that we are adding the extra mail_board functionality that it is not required right now for us :thinking:

luisDIXMIT commented on voip_oca/static/src/components/call_list/call_list.xml:
```diff
@@ -11,8 +11,8 @@
-                    <t t-esc="call.startDate" />
-                    <t t-if="call.duration"> • <t t-esc="call.duration" />
+                    <t t-esc="call.dateStr" />
+                    <t t-if="call.duration"> • <t t-esc="call.durationStr" />
```
maybe should be 't-if="call.durationStr"'

etobella replied:
no need, as durationStr is computed from duration

luisDIXMIT commented on voip_oca/static/src/services/voip_oca_service.esm.js:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    /* Search functions */
+    async searchPartners(_search = "", offset = 0, limit = 13) {
```
I think that limit parameter must be set to 0 in order to retrieve all contacts not only the first 13 and get access to the full contact list. WDYT?

etobella replied:
I prefer to keep it simple, otherwise the widget will not be able to handle it if the Database is too big... Also, we should limit the values in the widget if possible (not only on the backend connection)

luisDIXMIT replied:
Copy mate. I see your point of view

JoanSForgeFlow committed:
[IMP] voip_oca: add summary to activity list

etobella committed:
[FIX] voip_oca: Multiple improvements

Fix readme

Fixing tests

Simplify models

Normalize phone number

[IMP] voip_oca: Override activity_main_partner

pedrobaeza commented on voip_oca/models/mail_activity_mixin.py:
```diff
@@ -0,0 +1,25 @@
+        compute="_compute_activity_main_partner_id",
+    )
+
+    def _compute_activity_main_partner_id(self):
+        for record in self:
+            record.activity_main_partner_id = record.get_activity_main_partner_id()
+
+    def get_activity_main_partner_id(self):
+        if self._name == "res.partner":
+            return self
+        if "partner_id" in self._fields:
+            return self.partner_id
+        return self.env["res.partner"]
```
No need of 2 methods. For inheritance, just override the compute:

```suggestion
    def _compute_activity_main_partner_id(self):
        self.activity_main_partner_id = False
        if "partner_id" in self._fields:
            for record in self:
                record.activity_main_partner_id = record.partner_id
```

and then on res.partner:

```
    def _compute_activity_main_partner_id(self):
        for record in self:
            record.activity_main_partner_id = record
```

etobella replied:
It is done this way in order to leave a hook for other modules.

pedrobaeza replied:
As said, the hook can be done in the compute method in the mixin or in the specific model. No need of extra method.

etobella replied:
Yes, but it is easier to extend it this way. Also, there are examples of this behavior inside OCA and Odoo. I would prefer to keep it as it is

pedrobaeza replied:
I don't think so, as the stacktrace is longer, and the code is less readable being split and this method is only called in one place, which is the clear indicator that shouldn't be extracted.

etobella replied:
Well, it was done this way to have it done as it is right now (I forgot to split it correctly). With this,  we can change it by thread model. It is longer, yes, but it is cleaner too IMO. Also, we can find several examples in Odoo Core:

https://github.com/odoo/odoo/blob/c315f5af7bbe47d6fa73d0dd0ba31ab6f93cfd03/addons/mail/models/mail_thread.py#L429
https://github.com/odoo/odoo/blob/c315f5af7bbe47d6fa73d0dd0ba31ab6f93cfd03/addons/hr_recruitment/models/hr_job.py#L339

pedrobaeza replied:
Well, that examples are not exactly the same and Odoo core can't be sometimes the best example, but this is not so important to block the PR.

pedrobaeza commented on voip_oca/models/res_partner.py:
```diff
@@ -0,0 +1,32 @@
+        }
+
+    @api.model
+    def get_voip_contacts(self, _search, offset, limit):
```
As I see you are doing some kind of "API", please prefix all the methods with `voip` / `_voip` instead:

```suggestion
    def voip_get_contacts(self, _search, offset, limit):
```

pedrobaeza commented on voip_oca/models/res_users.py:
```diff
@@ -0,0 +1,47 @@
+            "voip_password",
+        ]
+
+    def _get_voip_info(self):
```
```suggestion
    def _voip_get_info(self):
```

pedrobaeza commented on voip_oca/models/voip_oca_call.py:
```diff
@@ -0,0 +1,115 @@
+
+
+class VoipOcaCall(models.Model):
+    _name = "voip.oca.call"
```
I don't think we need `oca` on the model (and if Odoo is calling this the same, it's not a problem, as both modules are incompatible between them and we shouldn't look for the compatibility IMO).

```suggestion
    _name = "voip.call"
```

pedrobaeza commented on voip_oca/models/voip_oca_pbx.py:
```diff
@@ -0,0 +1,18 @@
+
+
+class VoipOcaPbx(models.Model):
+    _name = "voip.oca.pbx"
```
```suggestion
    _name = "voip.pbx"
```

pedrobaeza commented on voip_oca/static/audio/calltone.mp3:
Why not putting this in OGG?

etobella replied:
Because I found it by MP3. I can try to find OGG

pedrobaeza commented on voip_oca/static/description/actions.png:
I think this should go into `static/src/images` and use relative image path in the .md

etobella replied:
I don't understand it. It is used as relative paths now :thinking:

pedrobaeza commented:
Please check CI.

luisDIXMIT reviewed (approved):
LGTM!!

pedrobaeza reviewed (approved):
According live tests.

/ocabot merge nobump
